File: String_manipulation/Reorganize_string.py

'''
TC: O(nlogn)

Use a max heap-> b/c most frequent item here
1. this would mean O (logn)
2. can replace sorting for us

3. To use maxheap in python, replace count with -count
when pop min, actually poopping the max here

start with the most freq char
Use a hashmap
a -> 3      b -> 2     c -> 2

a: 3      a: 2      a: 1
b: 1 ->   b: 0  ->  b: 0
c: 1      c: 1      c: 0

Each time a letter is used, use a diff character
so use a previous to keep track of pointer


'''
import heapq
from collections import Counter


h = heapq

'''
The example used here is 
{ a: 3, b: 2, c: 1}
'''
def reorganizeString(s: str) ->str:

    # this will create a map with char: count
    # aaabbc
    # dict = {a: 3, b: 2, c: 1}
    count = Counter(s)
    mheap = []

    # use this instead of list comprehension
    # for char, count in count.items():
    #     mheap.append([-count, char])
    mheap = [[-count, char] for char, count in count.items()]
    h.heapify(mheap)   # O(n)

    # store previous character we use so no reuse

    prev = None
    # add it to result
    res = ""

    while mheap or prev:
        if prev and not mheap:
            return ""

        # each time we want most freq char
        # except prev

        # count = the count -1 and then put it back in
        count, char = h.heappop(mheap)
        res += char

        # instead of - because the whole thing is negative
        count += 1

        # already sth exists
        # this is after the count has already been descreased
        if prev:
            # [2, a]
            h.heappush(mheap, prev)
            prev = None

            # if it's already 0, then we don't add anything here

        if count != 0:
            prev = [count, char]
    print(res)
    return res

str = "aab"
reorganizeString(str)

File: Coding_challenges/reorganize_string.kt
package Coding_Challenges

import java.util.*
import kotlin.collections.HashMap

/*
https://leetcode.com/problems/reorganize-string/
TC: O(nlogn)

Use a max heap-> b/c most frequent item here
1. this would mean O (logn)
2. can replace sorting for us

start with the most freq char
Use a hashmap

a -> 3      b -> 2     c -> 2

each time we take off a, also then take off b, since this is the next most
freq elemetn
a: 3      a: 2      a: 1
b: 1 ->   b: 0  ->  b: 0
c: 1      c: 1      c: 0

Each time a letter is used, use a diff character
so use a previous to keep track of pointer

*/
// Any 2 adj not the same
fun reorganizeString(s: String): String {
        // Step 1: create a map with count and freq
    // build this part first
    val counts: MutableMap<Char, Int> = HashMap()

    for(char in s){
        // This can be simplifed using getOrDefault()
        counts.apply {
            put(char, getOrDefault(char, 0) + 1)
        }
    }
    // Build this based on how many times the character has occured
    val maxHeap: Queue<Char> = PriorityQueue { a: Char, b: Char ->
        counts.get(a)!!- counts.get(b)!!
    }
    // This will sort automatically
    // counts.keys at this point will be
    maxHeap.addAll(counts.keys)
    var results = StringBuilder()
    while(maxHeap!=null){
        // this is the most freq char
        var cur = maxHeap.poll()
        // next most frequent
        var next = maxHeap.poll()

        // Add these 2 first
        results.append(cur)
        results.append(next)

        // update the count
        counts.put(cur, counts.get(cur)!! - 1)
        counts.put(next, counts.get(next)!! - 1)

        // Once they are popped
        // Check if any more current element and next element left
        if(counts.get(cur)!! > 0){
            maxHeap.add(cur)
        }
        if(counts.get(next)!! >0){
            maxHeap.add(next)
        }

        // If there is 1 character left and basically
        // nothing left to split it, then we just have to return
        // an empty string as expected in this example before
        // return the empty string here
        if(!maxHeap.isEmpty()){
            var last = maxHeap.remove()
            // once removed
            if(counts.get(last)!! > 1){

                return ""
            }
            results.append(last)
        }
        return results.toString()
    }

    var cur = maxHeap.remove()
    var next = maxHeap.remove()
    return "joseph"
}

fun main() {

//    reorganizeString("jooooseph")
}


fun reorganizeString3(s:String){
// build this part first
    val counts: MutableMap<Char, Int> = HashMap()

    for(char in s){
        // This can be simplifed using getOrDefault()
        counts.apply {
            put(char, getOrDefault(char, 0) + 1)
        }
    }
    // Based on the number of times this has happened

    val maxHeap: Queue<Char> = PriorityQueue { a: Char, b: Char ->
        counts.get(a)!!- counts.get(b)!!
    }

    maxHeap.addAll(counts.keys)

    var result = StringBuilder()
    while(!maxHeap.isEmpty()){
        
        val res =maxHeap.poll()

    }
}